router/src/dml_handlers/rpc_write.rs

```diff
@@ -72,23 +102,64 @@ pub enum RpcWriteError {
 #[derive(Debug)]
 pub struct RpcWrite<T, C = CircuitBreaker> {
     endpoints: Balancer<T, C>,
+
+    /// The total number of distinct upstream ingesters to write a single
+    /// request to.
+    ///
+    /// Invariant: after an upstream has ACKed a write, it MUST NOT be sent the
+    /// same write again. This minimises duplication within the same upstream,
+    /// for which no additional durability is realised.
+    ///
+    /// NOTE: it is NOT possible to eliminate duplication of write requests
+    /// being sent to the same upstream, as an upstream (or a middle-man proxy)
+    /// may NACK a write, having already buffered the data. When this request is
+    /// retried, the data will be duplicated.
+    n_copies: usize,
 }
 
 impl<T> RpcWrite<T> {
     /// Initialise a new [`RpcWrite`] that sends requests to an arbitrary
     /// downstream Ingester, using a round-robin strategy.
-    pub fn new<N>(endpoints: impl IntoIterator<Item = (T, N)>, metrics: &metric::Registry) -> Self
+    ///
+    /// If [`Some`], `replica_copies` specifies the number of additional
+    /// upstream ingesters that must receive and acknowledge the write for it to
+    /// be considered successful.
+    ///
+    /// # Panics
+    ///
+    /// It's invalid to configure `replica_copies` such that more ACKs are
+    /// needed than the number of `endpoints`; doing so will cause a panic.
+    pub fn new<N>(
+        endpoints: impl IntoIterator<Item = (T, N)>,
+        replica_copies: Option<NonZeroUsize>,
+        metrics: &metric::Registry,
+    ) -> Self
     where
         T: Send + Sync + Debug + 'static,
         N: Into<Arc<str>>,
     {
+        let endpoints = Balancer::new(
+            endpoints
+                .into_iter()
+                .map(|(client, name)| CircuitBreakingClient::new(client, name.into())),
+            Some(metrics),
+        );
+
+        // Map the "replication factor" into the total number of distinct data
+        // copies necessary to consider a write a success.
+        let n_copies = replica_copies.map(NonZeroUsize::get).unwrap_or(1);
+
+        // Assert this configuration is not impossible to satisfy.
+        assert!(
+            n_copies <= endpoints.len(),
+            "cannot configure more write copies ({n_copies}) than ingester \
+            endpoints ({count})",
+            count = endpoints.len(),
+        );
+
         Self {
-            endpoints: Balancer::new(
-                endpoints
-                    .into_iter()
-                    .map(|(client, name)| CircuitBreakingClient::new(client, name.into())),
-                Some(metrics),
-            ),
+            endpoints,
+            n_copies: replica_copies.map(NonZeroUsize::get).unwrap_or(1),
```

Comment:
Is there a reason this is recomputing `n_copies` rather than doing:

```suggestion
            n_copies,
```

?